### PR TITLE
Completed weekly assessment

### PR DESCRIPTION
**Pull request by sam-saenz**
# Submitting Week 4 Assessment

### ✅ Completed Files

Mark the following as confirmation that the files in this repository are complete.

- [ ✅ ] JavaScript Coding Challenges
- [ ✅ ] Ruby Coding Challenges
- [ ✅ ] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -15,37 +15,115 @@
...
+  for (let i = 0; i < array.length; i++) {
+    let randIndex = Math.floor(Math.random() * (array.length));
+    let temp = array[i]
+    array[i] = array[randIndex]
+    array[randIndex] = temp
+  }
+  return array
+}
+
+
```
Comment (chrisyhlee) [inline code-challenges.test.js:51]: Well done!

File: code-challenges.test.js [hunk 2]
```diff
@@ -15,37 +15,115 @@
...
+})
 
 // b) Create the function that makes the test pass.
 
+// I will make a function called tallyVotes that takes in an object as parameter. Then, I will access the objects upVotes key, and subtract the stored value from the value in the downVotes key, and return the result.
+
+const tallyVotes = (obj) => {
+  return obj.upVotes - obj.downVotes
+}
+
```
Comment (chrisyhlee) [inline code-challenges.test.js:74]: Very clean! Nice use of dot notation to access those values.

File: code-challenges.test.js [hunk 3]
```diff
@@ -15,37 +15,115 @@
...
+  let newArr = []
+  dataArrays.forEach(arr => {
+    arr.forEach(element => {
+      if(!newArr.includes(element)){
+        newArr.push(element)
+      }
+    });
+  })
+  return newArr
+}
```
Comment (chrisyhlee) [inline code-challenges.test.js:129]: I like that you went for the refactor! As for the spread operator, if you put the spread operator right before your dataArrays parameter `...dataArrays`, it will allow you to pass in any number of arguments, and will store them all into a new array, allowing you to have as many arguments as you want in your function! Very close! Nice job!

File: interview-questions.md [hunk 4]
```diff
@@ -8,42 +8,42 @@ Then, research the question to expand on your answer. Even if you feel you have
...
-Researched answer:
+Researched answer: A block is essentially the same thing as a method in Ruby, except that where a method will always belong to a specific object, blocks are just generic code that can be invoked without being attatched to an object. they are similar to functions from other coding languages.
 
 1. How do you extract a value in a Ruby hash?
 
-Your answer:
+Your answer: To extract a value from a Ruby hash, you must use the bracket notation along with the key associated with the desired value. This can be done to extract the value for logic, or for something like printing to the console.
 
-Researched answer:
+Researched answer: You can also use other methods like delete to moodify or remove values from a Ruby hash, and values can also be updated after referencing them.
```
Comment (chrisyhlee) [inline interview-questions.md:38]: Great elaboration on these interview answers!

File: ruby-challenges.rb [hunk 5]
```diff
@@ -3,32 +3,71 @@
...
+        if num.even?
+            "#{num} is even"
+        else
+            "#{num} is odd"
+        end
+    else
+        'Bad input'
+    end
+end
+
```
Comment (chrisyhlee) [inline ruby-challenges.rb:19]: Nice touch in handling edge cases!

File: ruby-challenges.rb [hunk 6]
```diff
@@ -3,32 +3,71 @@
...
 
 # -------------------2) Create a method that takes in a string and removes all the vowels from the string. Use the test variables provided. 
 # HINT: Check out this resource: https://ruby-doc.org/core-2.6/String.html#method-i-delete
 
+# Define a function removeVowels that takes in a string and uses .delete to remove all uppercase and lowercase vowels from the input.
+
+def removeVowels(string)
+    string.delete "aeiouAEIOU"
+end
+
```
Comment (chrisyhlee) [inline ruby-challenges.rb:40]: Clean, love it!

File: ruby-challenges.rb [hunk 7]
```diff
@@ -3,32 +3,71 @@
...
+# Define a function check_for_palindrome that takes in a string as an argument, then converts it to all lowercase using the .downcase function, checks if that is equal to the string after using the .reverse function, and returns the applicable string interpolated result.
+
+def check_for_palindrome string
+    if string.downcase == string.downcase.reverse
+        p "#{string} is a palindrome"
+    else
+        p "#{string} is not a palindrome"
+    end
+end
+
```
Comment (chrisyhlee) [inline ruby-challenges.rb:64]: Well done!

**Review by chrisyhlee — APPROVED** [review]
Sam, this was a great assessment, really good work! This assessment displays Advanced Beginner comprehension level. I would look over JavaScript 3 and see how you could refactor it to make it cleaner and more efficient. Overall though, excellent work, you are right on that border for Competent level, keep it up!
